File: javascript/bowling/bowling.js
//
// This is only a SKELETON file for the 'Bowling' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Bowling {
  constructor() {
    this.rolls = [];
    this._gameOver = false; // Internal flag to track game completion
  }

  /**
   * Record a roll in the game.
   * @param {number} pins The number of pins knocked down.
   */
  roll(pins) {
    if (this._gameOver) {
      throw new Error('Cannot roll after game is over');
    }
    if (pins < 0) {
      throw new Error('Negative roll is invalid');
    }
    if (pins > 10) {
      throw new Error('Pin count exceeds pins on the lane');
    }

    // Validate pin count based on current frame rules before adding the roll
    this._validateRollRules(pins);

    this.rolls.push(pins);

    // Check if this roll concludes the game
    this._checkAndUpdateGameOverState();
  }

  /**
   * Calculate the total score for the game.
   * @returns {number} The total score.
   */
  score() {
    if (!this._isGameComplete()) {
        // Check if the game has started at all
        if (this.rolls.length === 0) {
             throw new Error('Score cannot be taken until the end of the game');
        }
        // Check based on expected rolls vs actual rolls
        if (!this._isGameComplete()) { // Re-check for clarity and robustness
             throw new Error('Score cannot be taken until the end of the game');
        }
    }
    // Mark game as over definitively if scoring is allowed and calculation proceeds
    this._gameOver = true;


    let totalScore = 0;
    let rollIndex = 0;
    for (let frame = 1; frame <= 10; frame++) {
      if (this._isStrike(rollIndex)) {
        totalScore += 10 + this._strikeBonus(rollIndex);
        rollIndex++;
      } else if (this._isSpare(rollIndex)) {
        totalScore += 10 + this._spareBonus(rollIndex);
        rollIndex += 2;
      } else {
        totalScore += this._frameScore(rollIndex);
        rollIndex += 2;
      }
    }
    return totalScore;
  }

  // --- Private Helper Methods ---

  /** Checks if the current roll ('pins') is valid based on game rules and current state. */
  _validateRollRules(pins) {
    let currentFrame = 1;
    let rollIndex = 0;
    let firstRollInFrame = true;

    // Determine the frame number and roll context for the *next* roll
    while (currentFrame <= 10 && rollIndex < this.rolls.length) {
        if (this._isStrike(rollIndex)) {
            // Strike completes the frame in one roll (except frame 10)
            if (currentFrame === 10) {
                 // Need to know how many rolls already in 10th
                 const tenthFrameStartIndex = this._calculateFrameStartIndex(10);
                 const rollsInTenth = this.rolls.length - tenthFrameStartIndex;
                 if (rollsInTenth === 1) { // After 1st strike roll
                     firstRollInFrame = false; // Next is 2nd roll (1st bonus)
                 } else if (rollsInTenth === 2) { // After 2nd roll (1st bonus)
                     firstRollInFrame = false; // Next is 3rd roll (2nd bonus)
                 }
                 // Stay in frame 10 logic
                 rollIndex++; // Advance past the roll being considered
                 break; // Exit loop, handle frame 10 below
            } else {
                 // Strike in frames 1-9
                 rollIndex++;
                 currentFrame++;
                 firstRollInFrame = true;
            }
        } else {
            // Not a strike on the first roll
            if (rollIndex + 1 >= this.rolls.length) {
                // Game ends mid-frame (only one roll recorded)
                firstRollInFrame = false; // Next roll is the second roll
                rollIndex++;
                // currentFrame stays the same
                break; // Exit loop, handle current frame below
            } else {
                // Two rolls completed the frame
                const first = this.rolls[rollIndex];
                const second = this.rolls[rollIndex + 1];
                if (currentFrame === 10 && first + second === 10) { // Spare in 10th
                    const tenthFrameStartIndex = this._calculateFrameStartIndex(10);
                    const rollsInTenth = this.rolls.length - tenthFrameStartIndex;
                     if (rollsInTenth === 2) { // After spare completion
                         firstRollInFrame = false; // Next is 3rd roll (bonus)
                     }
                     // Stay in frame 10 logic
                     rollIndex += 2; // Advance past the rolls being considered
                     break; // Exit loop, handle frame 10 below
                } else {
                     // Open frame (frames 1-9 or 10)
                     rollIndex += 2;
                     currentFrame++;
                     firstRollInFrame = true;
                }
            }
        }
    }

    // Validate 'pins' based on the calculated state (currentFrame, firstRollInFrame)
    if (currentFrame <= 9) {
        if (!firstRollInFrame) {
            // Validating the second roll of frames 1-9
            const firstRollPins = this.rolls[this.rolls.length - 1]; // The actual previous roll
            if (firstRollPins + pins > 10) {
                throw new Error('Pin count exceeds pins on the lane');
            }
        }
        // If firstRollInFrame is true, any roll 0-10 is valid (pins > 10 already checked).
    } else if (currentFrame === 10) {
        const tenthFrameStartIndex = this._calculateFrameStartIndex(10);
        // If frame 10 hasn't started, tenthFrameStartIndex will be == this.rolls.length or -1 if rolls is empty
        if (tenthFrameStartIndex === -1 || tenthFrameStartIndex === this.rolls.length) {
             // This is the first roll of the 10th frame. Any 0-10 is valid.
             return;
        }

        const rollsInTenthSoFar = this.rolls.length - tenthFrameStartIndex;
        const firstRoll = this.rolls[tenthFrameStartIndex];
        const secondRoll = (rollsInTenthSoFar >= 1) ? this.rolls[tenthFrameStartIndex + 1] : undefined;

        if (rollsInTenthSoFar === 1) {
            // Validating the second roll of the 10th frame
            if (firstRoll !== 10 && firstRoll + pins > 10) {
                throw new Error('Pin count exceeds pins on the lane');
            }
        } else if (rollsInTenthSoFar === 2) {
            // Validating the third (bonus) roll of the 10th frame
            const isStrike = firstRoll === 10;
            // Ensure secondRoll is defined before checking spare
            const isSpare = !isStrike && secondRoll !== undefined && (firstRoll + secondRoll === 10);

            if (!isStrike && !isSpare) {
                // Should not be possible to roll a 3rd time on an open frame.
                // The main roll() check for _gameOver should prevent this.
                throw new Error('Cannot roll after open frame in 10th'); // Defensive
            }

            if (isStrike) {
                // Bonus roll after a strike
                // If the first bonus roll (secondRoll) was NOT a strike,
                // the sum of the two bonus rolls cannot exceed 10.
                if (secondRoll !== undefined && secondRoll !== 10 && secondRoll + pins > 10) {
                    throw new Error('Pin count exceeds pins on the lane');
                }
            }
            // If it was a spare, the bonus roll 'pins' (0-10) is valid.
            // If it was strike-strike, the bonus roll 'pins' (0-10) is valid.
        }
        // If rollsInTenthSoFar >= 3, the main roll() check should prevent rolling.
    }
    // If currentFrame > 10, the main roll() check should prevent rolling.
  }


  /** Checks if the game is over based on rolls and updates the internal flag. */
  _checkAndUpdateGameOverState() {
    const tenthFrameStartIndex = this._getFrameStartIndex(10);
    if (tenthFrameStartIndex === -1) return; // Game hasn't reached the 10th frame

    const rollsInTenth = this.rolls.length - tenthFrameStartIndex;
    if (rollsInTenth < 2) return; // 10th frame incomplete

    const firstRoll = this.rolls[tenthFrameStartIndex];
    // Ensure second roll exists before checking spare/open frame completion
    if (rollsInTenth >= 2) {
        const secondRoll = this.rolls[tenthFrameStartIndex + 1];
        const isStrike = firstRoll === 10;
        const isSpare = !isStrike && (firstRoll + secondRoll === 10);

        if (isStrike) {
            if (rollsInTenth === 3) this._gameOver = true;
        } else if (isSpare) {
            if (rollsInTenth === 3) this._gameOver = true;
        } else { // Open frame
            // Game ends after 2 rolls in an open 10th frame
            this._gameOver = true;
        }
    }
  }

   /** Determines if the game has the required number of rolls to be considered complete. */
   _isGameComplete() {
        let expectedRolls = 0;
        let tempRollIndex = 0;
        let frame = 1;
        while(frame <= 10) {
            if (tempRollIndex >= this.rolls.length) return false; // Not enough rolls recorded yet

            const first = this.rolls[tempRollIndex];
            if (first === 10) { // Strike
                expectedRolls++;
                tempRollIndex++;
                if (frame === 10) expectedRolls += 2; // Strike in 10th needs 2 bonus rolls
            } else { // Not a strike on the first ball
                expectedRolls += 2; // Expect two rolls for the frame
                 if (tempRollIndex + 1 >= this.rolls.length) return false; // Second roll missing
                 const second = this.rolls[tempRollIndex + 1];
                 tempRollIndex += 2;
                 if (frame === 10 && first + second === 10) {
                     expectedRolls++; // Spare in 10th needs 1 bonus roll
                 }
            }
            frame++;
        }
        // Game is complete if the number of rolls matches the expected number
        return this.rolls.length === expectedRolls;
   }


  /** Calculates the start index of a given frame (1-based). Returns -1 if frame not started. */
  _getFrameStartIndex(targetFrame) {
      return this._calculateFrameStartIndex(targetFrame);
  }

  /** Robust calculation of frame start index, used by validation and scoring. */
   _calculateFrameStartIndex(targetFrame) {
        let rollIndex = 0;
        let currentFrame = 1;
        while (currentFrame < targetFrame && rollIndex < this.rolls.length) {
            if (this._isStrike(rollIndex)) {
                rollIndex++;
            } else {
                 // Need two rolls to complete a non-strike frame
                 if (rollIndex + 1 >= this.rolls.length) return -1; // Frame incomplete
                rollIndex += 2;
            }
            currentFrame++;
        }
        // If we finished because we reached the target frame
        if (currentFrame === targetFrame) {
             // Ensure the calculated index is valid within the current rolls
             // (It's possible to calculate an index for a frame that *should* start but hasn't yet)
             if (rollIndex > this.rolls.length) return -1;
             return rollIndex;
        }
        // If we finished because we ran out of rolls before reaching the target frame
        return -1;
    }


  _isStrike(rollIndex) {
    return this.rolls[rollIndex] === 10;
  }

  _isSpare(rollIndex) {
    // Check that both rolls exist and sum to 10
    return rollIndex + 1 < this.rolls.length &&
           this.rolls[rollIndex] + this.rolls[rollIndex + 1] === 10;
  }

  _strikeBonus(rollIndex) {
    // Ensure the next two rolls exist for the bonus calculation
    if (rollIndex + 2 >= this.rolls.length) {
        // Handle cases where bonus rolls might be missing (e.g., strike in 9th, only one roll in 10th)
        // The score() check should prevent scoring incomplete games, but add safety.
        const bonus1 = this.rolls[rollIndex + 1] !== undefined ? this.rolls[rollIndex + 1] : 0;
        const bonus2 = this.rolls[rollIndex + 2] !== undefined ? this.rolls[rollIndex + 2] : 0;
        return bonus1 + bonus2;
    }
    return this.rolls[rollIndex + 1] + this.rolls[rollIndex + 2];
  }

  _spareBonus(rollIndex) {
    // Ensure the roll after the spare exists
    if (rollIndex + 2 >= this.rolls.length) {
        return 0; // No bonus if the next roll doesn't exist
    }
    return this.rolls[rollIndex + 2];
  }

  _frameScore(rollIndex) {
    // Ensure both rolls of the frame exist
    if (rollIndex + 1 >= this.rolls.length) {
        // Should only happen in an incomplete game, return score so far.
        return this.rolls[rollIndex] !== undefined ? this.rolls[rollIndex] : 0;
    }
    return this.rolls[rollIndex] + this.rolls[rollIndex + 1];
  }
}

File: python/bowling/bowling.py
class BowlingGame:
    def __init__(self):
        self.rolls = []

    def roll(self, pins):
        if pins < 0 or pins > 10:
            raise Exception("Invalid number of pins")

        # Determine current frame
        frame_index = 0
        i = 0
        while i < len(self.rolls) and frame_index < 9:
            if self.rolls[i] == 10:
                i += 1
            else:
                i += 2
            frame_index += 1

        # Validation for frames 1-9
        if frame_index < 9:
            # Find start index of current frame
            frame_start = 0
            frame = 0
            idx = 0
            while frame < frame_index and idx < len(self.rolls):
                if self.rolls[idx] == 10:
                    idx += 1
                else:
                    idx += 2
                frame += 1
            frame_start = idx

            # If this is the second roll in the current frame, validate sum
            if len(self.rolls) == frame_start + 1:
                first_roll = self.rolls[frame_start]
                if first_roll != 10 and first_roll + pins > 10:
                    raise Exception("Frame pin count exceeds 10")

            if self.is_game_over():
                raise Exception("Cannot roll after game is over")

        # Validation for 10th frame
        if frame_index == 9:
            tenth_frame = self.rolls[i:]
            if len(tenth_frame) == 0:
                pass  # first roll, always valid
            elif len(tenth_frame) == 1:
                first = tenth_frame[0]
                if first != 10 and first + pins > 10:
                    raise Exception("Bonus roll pin count exceeds 10")
            elif len(tenth_frame) == 2:
                first, second = tenth_frame[0], tenth_frame[1]
                if first == 10:
                    # strike in first roll
                    if second != 10 and second + pins > 10:
                        raise Exception("Bonus roll pin count exceeds 10")
                    if second != 10 and pins == 10:
                        raise Exception("Invalid strike after non-strike bonus roll")
                elif first + second == 10:
                    # spare in first two rolls, third roll unrestricted (max 10)
                    pass
                else:
                    # open frame, no third roll allowed
                    raise Exception("No bonus roll allowed")

            # Check if 10th frame is already over
            tenth_len = len(tenth_frame)
            if tenth_len == 3:
                raise Exception("Cannot roll after game is over")

        else:
            # Not in 10th frame, check if game is over before appending
            if self.is_game_over():
                raise Exception("Cannot roll after game is over")

        # Append only after validation
        self.rolls.append(pins)

    def score(self):
        if not self.rolls:
            raise Exception("Game not started")

        score = 0
        roll_index = 0
        for frame in range(10):
            if roll_index >= len(self.rolls):
                raise Exception("Incomplete game")

            if self.rolls[roll_index] == 10:
                if roll_index + 2 >= len(self.rolls):
                    raise Exception("Strike bonus rolls missing")
                score += 10 + self.rolls[roll_index + 1] + self.rolls[roll_index + 2]
                roll_index += 1
            elif roll_index + 1 < len(self.rolls):
                frame_score = self.rolls[roll_index] + self.rolls[roll_index + 1]
                if frame_score > 10:
                    raise Exception("Invalid frame score")
                if frame_score == 10:
                    if roll_index + 2 >= len(self.rolls):
                        raise Exception("Spare bonus roll missing")
                    score += 10 + self.rolls[roll_index + 2]
                else:
                    score += frame_score
                roll_index += 2
            else:
                raise Exception("Incomplete frame")

        if roll_index < len(self.rolls):
            # Parse frames explicitly to identify 10th frame and bonus rolls
            rolls = self.rolls
            frame = 0
            i = 0
            while frame < 9 and i < len(rolls):
                if rolls[i] == 10:
                    i += 1
                else:
                    i += 2
                frame += 1

            # Now at start of 10th frame
            tenth_frame = rolls[i:]
            if len(tenth_frame) < 2:
                raise Exception("Incomplete 10th frame")

            first = tenth_frame[0]
            second = tenth_frame[1] if len(tenth_frame) > 1 else None
            third = tenth_frame[2] if len(tenth_frame) > 2 else None

            if first == 10:
                # strike in first roll
                if second is None or third is None:
                    raise Exception("Strike bonus rolls missing")
                if len(tenth_frame) > 3:
                    raise Exception("Too many bonus rolls after strike in 10th frame")
            elif first + second == 10:
                # spare
                if third is None:
                    raise Exception("Spare bonus roll missing")
                if len(tenth_frame) > 3:
                    raise Exception("Too many rolls after spare in 10th frame")
            else:
                # open frame
                if len(tenth_frame) > 2:
                    raise Exception("Extra rolls after game over")

        return score

    def is_game_over(self):
        rolls = self.rolls
        frame = 0
        i = 0
        while frame < 10 and i < len(rolls):
            if rolls[i] == 10:
                i += 1
            else:
                if i + 1 >= len(rolls):
                    return False
                i += 2
            frame += 1

        if frame < 10:
            return False

        # 10th frame bonus logic
        tenth_frame = rolls[i - 2:]
        if len(tenth_frame) < 2:
            return False  # 10th frame incomplete

        first = tenth_frame[0]
        second = tenth_frame[1] if len(tenth_frame) > 1 else None
        third = tenth_frame[2] if len(tenth_frame) > 2 else None

        if first == 10:
            # strike in first roll
            if second is None or third is None:
                return False
            return True
        elif first + second == 10:
            # spare in first two rolls
            if third is None:
                return False
            return True
        else:
            # open frame, only two rolls allowed
            return True
